src/reconcile/rec_child.c

```diff
@@ -44,12 +44,20 @@ __rec_child_deleted(
 
     /*
      * If an earlier reconciliation chose to write the fast truncate information to the page, we
-     * should select it regardless of visibility unless it is visible globally visible. This is
-     * important as it is never ok to shift the on-disk value backwards.
+     * should select it regardless of visibility unless it is globally visible. This is important as
+     * it is never ok to shift the on-disk value backwards.
      */
     if (page_del->selected_for_write && !visible_all) {
         cmsp->del = *page_del;
         cmsp->state = WT_CHILD_PROXY;
+        /*
+         * We rely on the next reconciliation to remove the deleted entries if they become globally
+         * visible. Mark the page dirty for checkpoint to ensure we will trigger the next
+         * reconciliation on this page. Otherwise, the deleted pages may never be removed from disk.
+         * For eviction, the page is marked dirty the next time it is read into the memory.
+         */
+        if (!F_ISSET(r, WT_REC_EVICT))
+            r->leave_dirty = true;
```

Comment:
Yes, in 6.0. We do this as well. I think this works together with the change to mark the page dirty when reading it in memory. In summary, this ensures all the pages with deleted pages in memory are marked as dirty.

Reply:
I don't think this change should be included here. Now that version information is persisted to disk, it's safe to evict content from the cache. It isn't up to reconciliation to decide whether that was the best choice - that's a decision eviction makes.
I don't see a similar check in the 6.0 code, the only similar code I see is this:

```
    /*
     * Internal pages with deletes that aren't stable cannot be evicted, we don't have sufficient
     * information to restore the page's information if subsequently read (we wouldn't know which
     * transactions should see the original page and which should see the deleted page).
     */
    if (F_ISSET(r, WT_REC_EVICT))
        return (__wt_set_return(session, EBUSY));
```

Note that it applies only for evictions (not checkpoint reconciliations) which is the opposite of the proposed change here.

Reply:
```
   /*
     * Internal pages with deletes that aren't stable cannot be evicted, we don't have sufficient
     * information to restore the page's information if subsequently read (we wouldn't know which
     * transactions should see the original page and which should see the deleted page).
     */
    if (F_ISSET(r, WT_REC_EVICT))
        return (__wt_set_return(session, EBUSY));

    /* If the page cannot be marked clean. */
    r->leave_dirty = true;
```
6.0 set the leave_dirty flag to true, which is effectively the same as what I do.

Reply:
If we don't mark it as dirty, we only clean the deleted pages when this page is changed or evicted and read into memory again. Internal pages can take a long time to be evicted.

Reply:
@agorrod  After discussion with Hari, we think it is wrong in the mongodb flow to not mark the page dirty in checkpoint. If checkpoint clean the page, and the page deleted information is not globally visible, the next eviction will just evict the page without doing reconciliation again, which is wrong in the mongodb flow. Also it will delay the time to clean the deleted pages until the next time the page is evicted and then read into memory.

Reply:
My understanding is that even though in MongoDB oplog case, the truncated oplog is never accessed again, keeping the page dirty will result in rewriting the internal page again and again for every checkpoint till it becomes globally visible. Since the page is rewritten again, the previous disk image will be reclaimed and there will not be an increase in disk space.
If we don't mark the page dirty, we might lose the page_del info, if the eviction evicts this page as checkpoint marked it clean, which might not be a problem in MongoDB oplog case, but I think this change is correct.
We need to do testing by flipping the standalone flag, to test more MongoDB scenarios.